ledger/sandbox/src/main/scala/com/digitalasset/platform/index/TransactionConversion.scala

```diff
@@ -3,206 +3,240 @@
 
   // Remove root nodes that have empty witnesses and put their children in their place as roots.
   // Do this while there are roots with no witnesses.
   @tailrec
   private def removeInvisibleRoots(
-      eventsById: Map[domain.EventId, CreateOrExerciseEvent],
-      rootEventIds: List[domain.EventId])
-    : (Map[domain.EventId, CreateOrExerciseEvent], List[domain.EventId]) = {
+      eventsById: Map[String, TreeEvent],
+      rootEventIds: Seq[String]): (Map[String, TreeEvent], Seq[String]) = {
 
     val result =
-      rootEventIds.foldRight(InvisibleRootRemovalState(rootsWereReplaced = false, eventsById, Nil)) {
+      rootEventIds.foldRight(
+        InvisibleRootRemovalState(rootsWereReplaced = false, eventsById, Seq.empty)) {
         case (eventId, InvisibleRootRemovalState(hasInvisibleRoot, filteredEvents, newRoots)) =>
           val event = eventsById
             .getOrElse(
               eventId,
               throw new IllegalArgumentException(
                 s"Root event id $eventId is not present among transaction nodes ${eventsById.keySet}"))
           if (event.witnessParties.nonEmpty)
-            InvisibleRootRemovalState(hasInvisibleRoot, filteredEvents, eventId :: newRoots)
+            InvisibleRootRemovalState(hasInvisibleRoot, filteredEvents, eventId +: newRoots)
           else
             InvisibleRootRemovalState(
               rootsWereReplaced = true,
               filteredEvents - eventId,
-              event.children ::: newRoots)
+              event.childEventIds ++ newRoots)
       }
     if (result.rootsWereReplaced)
       removeInvisibleRoots(result.eventsById, result.rootEventIds)
     else (result.eventsById, result.rootEventIds)
   }
 
-  private def lfCreateToDomain(
-      eventId: domain.EventId,
-      create: CreateEvent[Lf.AbsoluteContractId],
+  private def lfCreateToApiCreate(
+      eventId: String,
+      create: CreateEvent[AbsoluteContractId],
       includeParentWitnesses: Boolean,
-  ): domain.Event.CreatedEvent = {
-    domain.Event.CreatedEvent(
+      verbose: Boolean) = {
+    CreatedEvent(
       eventId,
-      domain.ContractId(create.contractId.coid),
-      create.templateId,
-      create.argument.value match {
-        case rec @ Lf.ValueRecord(_, _) => rec
-        case _ => throw new RuntimeException(s"Value is not an record.")
-      },
-      if (includeParentWitnesses) create.witnesses
-      else create.stakeholders,
-      create.signatories,
-      create.observers,
-      create.agreementText,
-      create.contractKey.map(_.key.value)
+      create.contractId.coid,
+      Some(LfEngineToApi.toApiIdentifier(create.templateId)),
+      create.contractKey.map(
+        ck =>
+          LfEngineToApi.assertOrRuntimeEx(
+            "translating the contract key",
+            LfEngineToApi
+              .lfValueToApiValue(verbose, ck.key.value))),
+      Some(
+        LfEngineToApi
+          .lfValueToApiRecord(verbose = true, create.argument.value match {
```

Comment:
Should this not use the `verbose` parameter?

Reply:
Ah, good catch, thanks.

Reply:
Addressed in https://github.com/digital-asset/daml/pull/4763/commits/084f83f6a7e9794c09cd508e4c4c0898039e0bea